GNU calling convention broken

This "[Improve const correctness for P11KitUri](https://github.com/p11-glue/p11-kit/commit/e42dcf5283a5537c196147c9a2468ee537b9da7b)" commit appears to break compilation of `gnutls` due to use of `CRYPTOKI_GNU`.

This results in error messages such as:

```
/usr/local/Cellar/p11-kit/0.23.11/include/p11-kit-1/p11-kit/uri.h:134:68: error: unknown type name 'CK_ATTRIBUTE'; did you mean 'CK_ATTRIBUTE_PTR'?
                                                             const CK_ATTRIBUTE *attrs,
                                                                   ^~~~~~~~~~~~
                                                                   CK_ATTRIBUTE_PTR
/usr/local/Cellar/p11-kit/0.23.11/include/p11-kit-1/p11-kit/uri.h:85:30: note: 'CK_ATTRIBUTE_PTR' declared here
typedef struct ck_attribute *CK_ATTRIBUTE_PTR;
```
This was encountered while trying to build `gnutls-3.5.18.tar.xz` via Homebrew on a Mac.

[Apologies for the brevity of report I'm using a non-GitHub supported browser and have bypassed the immediate issue by building `gnutls` without `p11-kit` but hoped this level of detail was better than no report at all. :) ]

--- a/.gitignore
+++ b/.gitignore
@@ -99,6 +99,7 @@ x86_64-w64-mingw32
 /p11-kit/pkcs11.conf.example
 /p11-kit/p11-kit-remote@.service
 /p11-kit/virtual-fixed.c
+/p11-kit/*.gnu.c
 
 /p11-kit-remote
 /p11-kit-server

--- a/p11-kit/Makefile.am
+++ b/p11-kit/Makefile.am
@@ -104,6 +104,7 @@ noinst_LTLIBRARIES += \
 
 check_LTLIBRARIES += \
 	libp11-kit-testable.la \
+	libp11-kit-pkcs11-gnu.la \
 	$(NULL)
 
 libp11_kit_internal_la_LDFLAGS = -no-undefined
@@ -116,6 +117,37 @@ libp11_kit_testable_la_SOURCES = \
 	$(libp11_kit_la_SOURCES) \
 	$(NULL)
 libp11_kit_testable_la_LIBADD = $(COMMON_LIBS)
+
+pkcs11_gnu_sources = \
+	p11-kit/uri.gnu.c \
+	p11-kit/iter.gnu.c \
+	p11-kit/pin.gnu.c \
+	$(NULL)
+
+BUILT_SOURCES += $(pkcs11_gnu_sources)
+CLEANFILES += $(pkcs11_gnu_sources)
+
+p11-kit/uri.gnu.c: p11-kit/uri.h
+p11-kit/iter.gnu.c: p11-kit/iter.h
+p11-kit/pin.gnu.c: p11-kit/pin.h
+
+SUFFIXES = .h .gnu.c
+.h.gnu.c:
+	$(AM_V_GEN) src=$<; dst=$@; rm -f $@-t $@ && \
+	{ echo '/* DO NOT EDIT! GENERATED AUTOMATICALLY! */'; \
+	  echo; \
+	  echo '#include <stdlib.h>'; \
+	  echo '#include "p11-kit/p11-kit.h"'; \
+	  cat $$src; \
+	  echo "void *$${dst}_funcs[] = {" | sed 's/[^][ *a-z0-9_={]/_/g'; \
+	  sed -n -e '/^typedef/d' -e 's/.* \(p11_kit_[^ ]*\) *(.*/	\1,/p' $<; \
+	  echo '};'; \
+	} > $@-t && \
+	mv -f $@-t $@
+
+libp11_kit_pkcs11_gnu_la_LDFLAGS = -no-undefined
+libp11_kit_pkcs11_gnu_la_SOURCES = $(pkcs11_gnu_sources)
+libp11_kit_pkcs11_gnu_la_CFLAGS = -DCRYPTOKI_GNU=1 -DP11_KIT_FUTURE_UNSTABLE_API=1
 
 if OS_WIN32
 

--- a/p11-kit/iter.h
+++ b/p11-kit/iter.h
@@ -45,6 +45,16 @@ extern "C" {
 
 #ifdef P11_KIT_FUTURE_UNSTABLE_API
 
+/*
+ * If the caller is using the PKCS#11 GNU calling convention, then we cater
+ * to that here.
+ */
+#ifdef CRYPTOKI_GNU
+typedef unsigned char CK_BBOOL;
+typedef ck_object_handle_t CK_OBJECT_HANDLE;
+typedef ck_session_handle_t CK_SESSION_HANDLE;
+#endif
+
 typedef struct p11_kit_iter P11KitIter;
 typedef P11KitIter p11_kit_iter;
 

--- a/p11-kit/uri.h
+++ b/p11-kit/uri.h
@@ -79,11 +79,15 @@ typedef enum {
  * to that here.
  */
 #ifdef CRYPTOKI_GNU
+typedef struct ck_info CK_INFO;
 typedef struct ck_info *CK_INFO_PTR;
+typedef struct ck_token_info CK_TOKEN_INFO;
 typedef struct ck_token_info *CK_TOKEN_INFO_PTR;
 typedef ck_attribute_type_t CK_ATTRIBUTE_TYPE;
+typedef struct ck_attribute CK_ATTRIBUTE;
 typedef struct ck_attribute *CK_ATTRIBUTE_PTR;
 typedef unsigned long int CK_ULONG;
+typedef struct ck_slot_info CK_SLOT_INFO;
 typedef struct ck_slot_info *CK_SLOT_INFO_PTR;
 typedef ck_slot_id_t CK_SLOT_ID;
 typedef P11KitUriType p11_kit_uri_type_t;

